src/azul/bigquery_reservation.py

```diff
@@ -0,0 +1,141 @@
+                                      reservation=reservation_name,
+                                      assignment=assignment_name)
+        else:
+            log.info('Slot commitment already active')
+        return resources
+
+    def delete_slots(self, resources: ResourceNames) -> None:
+        self._client.delete_assignment(name=resources.assignment)
+        self._client.delete_reservation(name=resources.reservation)
+        self._client.delete_capacity_commitment(name=resources.commitment)
+        log.info('Deleted resources: %r', resources)
+
+    def get_resource_names(self) -> Optional[ResourceNames]:
+        args = {}
+        for resource_type, method, parent in [
+            ('commitment', self._client.list_capacity_commitments, self._reservation_parent_path),
+            ('reservation', self._client.list_reservations, self._reservation_parent_path),
+            ('assignment', self._client.list_assignments, self._reservation_parent_path + '/reservations/-')
+        ]:
+            resource_name = self._single_name(resource_type, method(parent=parent))
+            if resource_name is not None:
+                args[resource_type] = resource_name
+        try:
+            return ResourceNames(**args)
+        except TypeError:
+            require(not args, f'BigQuery slot commitment state is inconsistent--'
+                              f'some, but not all resource are missing: {args}')
+            return None
+
+    def _single_name(self, resource_type: str, resources) -> Optional[str]:
+        try:
+            return one((r.name for r in resources),
+                       too_long=RequirementError(f'Too many {resource_type}s in '
+                                                 f'path {self._reservation_parent_path} '
+                                                 f'(should be 0 or 1)'))
+        except ValueError:
+            return None
+
+    def _purchase_commitment(self) -> CapacityCommitment:
+        commitment = CapacityCommitment(dict(
+            slot_count=self.slots,
+            plan=CapacityCommitment.CommitmentPlan.FLEX)
+        )
+        commitment = self._client.create_capacity_commitment(
+            capacity_commitment=commitment,
+            parent=self._reservation_parent_path
+        )
+        log.info('Purchased %d BigQuery slots, commitment ID: %r',
+                 commitment.slot_count, commitment.name)
+        return commitment
+
+    def _create_reservation(self) -> Reservation:
+        reservation = Reservation(dict(
+            slot_capacity=self.slots,
+            ignore_idle_slots=False
+        ))
+        reservation = self._client.create_reservation(
+            reservation=reservation,
+            reservation_id=self._reservation_id,
+            parent=self._reservation_parent_path,
+        )
+        log.info('Reserving %d BigQuery slots, reservation ID: %r',
+                 reservation.slot_capacity, reservation.name)
```

Comment:
I'm a bit confused. Is this a name or an ID? What do these `.name` values look like?

Reply:
```
{'capacity_commitment_name': 'projects/platform-hca-dev/locations/US/capacityCommitments/17333357519783113156', 'reservation_name': 'projects/platform-hca-dev/locations/US/reservations/azul-reindex', 'assignment_name': 'projects/platform-hca-dev/locations/US/reservations/azul-reindex/assignments/16113151697275936322'}

Reply:
I called the variables `*_name` because the BigQuery reservation python API exposes them via `.name` attributes. But as you can see, "ID" seems like a better description of their contents.

Reply:
Use "name" consistently in favor of "ID".